src/Space.ts

```diff
@@ -1,119 +1,154 @@
+    const leaverIndex = this.leavers.findIndex(({ clientId }) => clientId === message.clientId);
 
-    if (!member) {
-      this.createMember(implicitClientId, isConnected, data || {});
-    } else {
-      member.isConnected = isConnected;
-      if (data) {
-        // Member data is completely overridden, except lastEventTimestamp which is updated
-        member.data = {
-          ...data,
-          lastEventTimestamp: new Date(),
-        };
-      }
+    if (message.action === 'leave' && leaverIndex < 0) {
+      this.addLeaver(message);
+    } else if (message.action === 'leave' && leaverIndex >= 0) {
+      this.resetLeaver(leaverIndex);
+      this.addLeaver(message);
+    } else if (leaverIndex >= 0) {
+      this.resetLeaver(leaverIndex);
+      this.removeLeaver(leaverIndex);
     }
 
-    const memberUpdateEvent = new MemberUpdateEvent(this.members);
-    this.dispatchEvent(memberUpdateEvent);
+    return this.members;
   }
 
-  enter(data: unknown) {
-    if (!data || typeof data !== 'object') {
-      return;
-    }
+  private subscribeToPresence() {
+    this.channel.presence.subscribe((message) => {
+      this.dispatchEvent(new SpaceMembersUpdateEvent(message));
+    });
+  }
 
-    const clientId = this.client.auth.clientId || undefined;
+  async enter(data?: unknown) {
     const presence = this.channel.presence;
 
-    // TODO: Discuss if we actually want change this behaviour in contrast to presence (enter becomes an update)
-    return presence.get({ clientId }).then(
-      (presenceMessages) =>
-        new Promise((resolve, reject) => {
-          if (presenceMessages && presenceMessages.length === 1) {
-            reject(new Error(ERROR_CLIENT_ALREADY_ENTERED));
-          }
-
-          this.syncMembers();
-          this.subscribeToPresenceEvents();
-          resolve(presence.enter(JSON.stringify(data)));
-        })
-    );
+    await presence.enter(data);
+    const presenceMessages = await presence.get();
+    this.members = this.mapPresenceMembersToSpaceMembers(presenceMessages);
+
+    return this.members;
   }
 
   leave(data?: unknown) {
-    const presence = this.channel.presence;
-    return presence.leave(data);
+    return this.channel.presence.leave(data);
   }
-}
 
-type SpaceMember = {
-  clientId: string;
-  isConnected: boolean;
-  data: { [key: string]: any };
-};
+  on(spaceEvent: SpaceEvents, callback) {
+    if (spaceEvent === 'membersUpdate') {
+      this.subscribeToPresence();
+      this.addEventListener('membersUpdate', (event: SpaceMembersUpdateEvent) => {
+        if (!event.message) return;
+        // By default, we only return data about other connected clients, not the whole set
+        if (event.message.clientId === this.clientId) return;
+
+        this.updateMembersFromPresenceMessage(event.message);
+        callback(this.members);
+      });
+    } else {
+      // TODO: align with ably-js error policy here
+      throw new Error(`Event "${spaceEvent}" is unsupported`);
```

Comment:
Maybe `42200: Unprocessable Entity`? @dpiatek

Reply:
https://github.com/ably/ably-common/blob/main/protocol/errors.json
Found the big list of errors

Reply:
Although I want us to align with Ably errors (as noted here https://github.com/ably-labs/spaces/pull/7#discussion_r1102629747) I'll leave this as it is for now. It seems wrong to just mix and match errors—we should have informative errors that let people look up an FAQ to solve their issue. I'm not sure however yet how to correctly extend what we have in ably-common